### PR TITLE
feat: Introduce ux CLI. Add optimize-images sub command.

### PR DESCRIPTION
**Pull request by mattwagl**
BREAKING CHANGE: Move verify-links into sub command of ux CLI.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -38,24 +38,34 @@ The integration tests include a [Next.js sample application](test/shared/sampleA
 $ npm run start-sample-application
 ```
 
+## Optimizing images
+
+To optimize JPG, PNG, and SVG images of a website, use the following command:
```
Comment (goloroden) [inline README.md]:
```suggestion
To optimize `.jpg`, `.png`, and `.svg` images of a website, use the following command:
```

File: README.md [hunk 2]
```diff
@@ -38,24 +38,34 @@ The integration tests include a [Next.js sample application](test/shared/sampleA
...
+
+```shell
+$ npx ux optimize-images --source <path> --destination <path>
+```
+
+The command recursively copies all images from the source folder to the destination folder and optimizes them using a reasonable preset.
```
Comment (goloroden) [inline README.md]:
```suggestion
This command recursively copies all images from the source directory to the destination directory and optimizes them using a reasonable preset.
```

File: lib/cli/optimizeImages/optimizeImages.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+}): Promise<void> {
+  const queue = new PQueue({ concurrency: os.cpus().length });
+
+  const imagePaths = await globby(source, {
+    expandDirectories: {
+      extensions: supportedExtensions
+    }
+  });
+
+  for (const sourcePath of imagePaths) {
```
Comment (goloroden) [inline lib/cli/optimizeImages/optimizeImages.ts]: Shouldn't this be `imagePath`?

```suggestion
  for (const imagePath of imagePaths) {
```

File: lib/cli/optimizeImages/optimizeImages.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    const relativeSourcePath = path.relative(source, sourcePath);
```
Comment (goloroden) [inline lib/cli/optimizeImages/optimizeImages.ts]: And, according to that, this should be `relativeImagePath`.

```suggestion
    const relativeImagePath = path.relative(source, imagePath);
```

File: lib/cli/optimizeImages/optimizeImages.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    const extensionName = path.extname(sourcePath).toLowerCase().replace('.', '') as SupportedExtensionNames;
+    const destinationPath = path.join(destination, relativeSourcePath);
+    const optimizeImage = optimizerForExtension[extensionName];
+
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    queue.add(async (): Promise<void> => {
+      buntstift.verbose(`Optimizing ${sourcePath}...`);
```
Comment (goloroden) [inline lib/cli/optimizeImages/optimizeImages.ts]:
```suggestion
      buntstift.verbose(`Optimizing '${sourcePath}'...`);
```

File: lib/cli/optimizeImages/optimizeImages.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+      await ensureFile(destinationPath);
+      await optimizeImage({ source: sourcePath, destination: destinationPath, verbose });
+
+      buntstift.verbose(`Optimized ${sourcePath}.`);
```
Comment (goloroden) [inline lib/cli/optimizeImages/optimizeImages.ts]:
```suggestion
      buntstift.verbose(`Optimized '${sourcePath}'.`);
```

File: lib/cli/optimizeImages/optimizeImagesCommand.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
+import { buntstift } from 'buntstift';
+import { Command } from 'command-line-interface';
+import { optimizeImages } from './optimizeImages';
+import { OptimizeImagesOptions } from './OptimizeImagesOptions';
+
+const optimizeImagesCommand = function (): Command<OptimizeImagesOptions> {
+  return {
+    name: 'optimize-images',
+    description: 'Optimizes the size of JPG, PNG, SVG images.',
```
Comment (goloroden) [inline lib/cli/optimizeImages/optimizeImagesCommand.ts]:
```suggestion
    description: 'Optimizes the size of .jpg, .png, and .svg images.',
```

File: lib/cli/optimizeImages/optimizeImagesCommand.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+    optionDefinitions: [
+      {
+        name: 'source',
+        alias: 's',
+        description: 'set the source folder',
+        parameterName: 'path',
+        type: 'string',
+        isRequired: true,
+        defaultOption: true
```
Comment (goloroden) [inline lib/cli/optimizeImages/optimizeImagesCommand.ts]: Why is this parameter the default option? Is this by intention, or a copy / paste error? If I have to copy from A to B, I wouldn't expect A to be the last parameter.

File: lib/cli/optimizeImages/optimizeJpg.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+  source: string;
+  destination: string;
+  verbose: boolean;
+}): Promise<void> {
+  const cwd = process.cwd();
+
+  await new Promise((resolve, reject): void => {
+    shell.exec(`${mozjpeg} -progressive -quality 70 -outfile "${destination}"  "${source}"`, { cwd, silent: !verbose }, (code): void => {
+      if (code !== 0) {
+        return reject(new Error(`Failed to optimize ${source}.`));
```
Comment (goloroden) [inline lib/cli/optimizeImages/optimizeJpg.ts]:
```suggestion
        return reject(new Error(`Failed to optimize '${source}'.`));
```

File: lib/cli/optimizeImages/optimizeJpg.ts [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+  verbose
+}: {
+  source: string;
+  destination: string;
+  verbose: boolean;
+}): Promise<void> {
+  const cwd = process.cwd();
+
+  await new Promise((resolve, reject): void => {
+    shell.exec(`${mozjpeg} -progressive -quality 70 -outfile "${destination}"  "${source}"`, { cwd, silent: !verbose }, (code): void => {
```
Comment (goloroden) [inline lib/cli/optimizeImages/optimizeJpg.ts]:
```suggestion
    shell.exec(`${mozjpeg} -progressive -quality 70 -outfile '${destination}'  '${source}'`, { cwd, silent: !verbose }, (code): void => {
```

File: lib/cli/optimizeImages/optimizePng.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+  verbose
+}: {
+  source: string;
+  destination: string;
+  verbose: boolean;
+}): Promise<void> {
+  const cwd = process.cwd();
+
+  await new Promise((resolve, reject): void => {
+    shell.exec(`${pngquant} --force --quality=80-90 --output "${destination}"  "${source}"`, { cwd, silent: !verbose }, (code): void => {
```
Comment (goloroden) [inline lib/cli/optimizeImages/optimizePng.ts]:
```suggestion
    shell.exec(`${pngquant} --force --quality=80-90 --output '${destination}' '${source}'`, { cwd, silent: !verbose }, (code): void => {
```

File: lib/cli/optimizeImages/optimizePng.ts [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+      if (code !== 0) {
+        return reject(new Error(`Failed to optimize ${source}.`));
```
Comment (goloroden) [inline lib/cli/optimizeImages/optimizePng.ts]:
```suggestion
        return reject(new Error(`Failed to optimize '${source}'.`));
```

File: lib/cli/optimizeImages/optimizeSvg.ts [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+  verbose
+}: {
+  source: string;
+  destination: string;
+  verbose: boolean;
+}): Promise<void> {
+  const cwd = process.cwd();
+
+  await new Promise((resolve, reject): void => {
+    shell.exec(`npx svgo ${source} -o ${destination}`, { cwd, silent: !verbose }, (code): void => {
```
Comment (goloroden) [inline lib/cli/optimizeImages/optimizeSvg.ts]:
```suggestion
    shell.exec(`npx svgo '${source}' -o '${destination}'`, { cwd, silent: !verbose }, (code): void => {
```

File: lib/cli/optimizeImages/optimizeSvg.ts [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+      if (code !== 0) {
+        return reject(new Error(`Failed to optimize ${source}.`));
```
Comment (goloroden) [inline lib/cli/optimizeImages/optimizeSvg.ts]:
```suggestion
        return reject(new Error(`Failed to optimize '${source}'.`));
```

File: test/cli/optimizeImages/optimizeImagesTests.ts [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+import path from 'path';
+
+const getFileSize = async function (filePath: string): Promise<number> {
+  const { size } = await fs.stat(filePath);
+
+  return size;
+};
+
+suite('optimizeImages', (): void => {
+  test('copies images from source to destination folder and optimizes their file size.', async (): Promise<void> => {
```
Comment (goloroden) [inline test/cli/optimizeImages/optimizeImagesTests.ts]:
```suggestion
  test('copies images from the source to the destination directory and optimizes their file sizes.', async (): Promise<void> => {
```

File: test/cli/optimizeImages/optimizeImagesTests.ts [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+
+    const optimizedSizes = {
+      jpg: await getFileSize(path.join(destinationDirectory, 'flat-images', 'image.jpg')),
+      png: await getFileSize(path.join(destinationDirectory, 'flat-images', 'image.png')),
+      svg: await getFileSize(path.join(destinationDirectory, 'flat-images', 'image.svg'))
+    };
+
+    assert.that(optimizedSizes.jpg < originalSizes.jpg).is.true();
+    assert.that(optimizedSizes.png < originalSizes.png).is.true();
+    assert.that(optimizedSizes.svg < originalSizes.svg).is.true();
```
Comment (goloroden) [inline test/cli/optimizeImages/optimizeImagesTests.ts]: This should be rewritten using `assert.that(…).is.lessThan(…)`.

File: test/cli/optimizeImages/optimizeImagesTests.ts [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+    const optimizedSizes = {
+      jpg: await getFileSize(path.join(destinationDirectory, 'nested-images', 'jpg', 'image.jpg')),
+      png: await getFileSize(path.join(destinationDirectory, 'nested-images', 'png', 'image.png')),
+      svg: await getFileSize(path.join(destinationDirectory, 'nested-images', 'svg', 'image.svg'))
+    };
+
+    assert.that(optimizedSizes.jpg < originalSizes.jpg).is.true();
+    assert.that(optimizedSizes.png < originalSizes.png).is.true();
+    assert.that(optimizedSizes.svg < originalSizes.svg).is.true();
```
Comment (goloroden) [inline test/cli/optimizeImages/optimizeImagesTests.ts]: This should be rewritten using `assert.that(…).is.lessThan(…)`.

File: test/styleguide/brokenUrlsTests.ts [hunk 18]
```diff
@@ -39,7 +39,7 @@ suite('styleguide', (): void => {
 
     await new Promise((resolve, reject): void => {
       shell.exec(
-        `npx ts-node ${cliPath} ${styleguideUrl} --verbose --ignore '${ignoreUrls}'`,
+        `npx ts-node ${cliPath} --verbose verify-links --url ${styleguideUrl} --ignore '${ignoreUrls}'`,
```
Comment (goloroden) [inline test/styleguide/brokenUrlsTests.ts]:
```suggestion
        `npx ts-node '${cliPath}' --verbose verify-links --url '${styleguideUrl}' --ignore '${ignoreUrls}'`,
```

File: lib/cli/optimizeImages/optimizeImagesCommand.ts (hunk 8 above)
Comment (mattwagl) [reply lib/cli/optimizeImages/optimizeImagesCommand.ts]: I wasn't sure about the semantics here. So thanks for the catch. Will remove the default option.

File: lib/cli/optimizeImages/optimizeImagesCommand.ts [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+const optimizeImagesCommand = function (): Command<OptimizeImagesOptions> {
+  return {
+    name: 'optimize-images',
+    description: 'Optimizes the size of .jpg, .png, and .svg images.',
+
+    optionDefinitions: [
+      {
+        name: 'source',
+        alias: 's',
+        description: 'set the source folder',
```
Comment (goloroden) [inline lib/cli/optimizeImages/optimizeImagesCommand.ts]:
```suggestion
        description: 'set the source directory',
```

File: lib/cli/optimizeImages/optimizeImagesCommand.ts [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+        parameterName: 'path',
+        type: 'string',
+        isRequired: true
+      },
+      {
+        name: 'destination',
+        alias: 'd',
+        description: 'set the destination folder',
```
Comment (goloroden) [inline lib/cli/optimizeImages/optimizeImagesCommand.ts]:
```suggestion
        description: 'set the destination directory',
```

File: lib/cli/optimizeImages/optimizeImagesCommand.ts (hunk 19 above)
Comment (goloroden) [reply lib/cli/optimizeImages/optimizeImagesCommand.ts]: This one has not yet been applied.

**Pull request merged**